packages/event-handler/src/rest/middleware/cors.ts

```diff
@@ -0,0 +1,219 @@
+  }
+  
+  if (typeof originConfig === 'string') {
+    return originConfig;
+  }
+  
+  return DEFAULT_CORS_OPTIONS.origin as string;
+}
+
+/**
+ * Handles preflight OPTIONS requests
+ */
+function handlePreflight(config: ResolvedCorsConfig, reqCtx: RequestContext): Response {
+  const { request } = reqCtx;
+  const requestOrigin = request.headers.get('Origin');
+  const resolvedOrigin = resolveOrigin(config.origin, requestOrigin, reqCtx);
+
+  const headers = new Headers();
+  
+  if (resolvedOrigin) {
+    headers.set('Access-Control-Allow-Origin', resolvedOrigin);
+  }
+  
+  if (config.allowMethods.length > 0) {
+    headers.set('Access-Control-Allow-Methods', config.allowMethods.join(', '));
+  }
+  
+  if (config.allowHeaders.length > 0) {
+    headers.set('Access-Control-Allow-Headers', config.allowHeaders.join(', '));
+  }
+  
+  if (config.credentials) {
+    headers.set('Access-Control-Allow-Credentials', 'true');
+  }
+  
+  if (config.maxAge !== undefined) {
+    headers.set('Access-Control-Max-Age', config.maxAge.toString());
+  }
+
+  return new Response(null, {
+    status: HttpErrorCodes.NO_CONTENT, // 204
+    headers,
+  });
+}
+
+/**
+ * Adds CORS headers to regular requests
+ */
+function addCorsHeaders(config: ResolvedCorsConfig, reqCtx: RequestContext): void {
+  const { request, res } = reqCtx;
+  const requestOrigin = request.headers.get('Origin');
+  const resolvedOrigin = resolveOrigin(config.origin, requestOrigin, reqCtx);
+
+  if (resolvedOrigin) {
+    res.headers.set('Access-Control-Allow-Origin', resolvedOrigin);
+  }
+  
+  if (config.exposeHeaders.length > 0) {
+    res.headers.set('Access-Control-Expose-Headers', config.exposeHeaders.join(', '));
+  }
+  
+  if (config.credentials) {
+    res.headers.set('Access-Control-Allow-Credentials', 'true');
+  }
+}
+
+/**
+ * Creates a CORS middleware that adds appropriate CORS headers to responses
+ * and handles OPTIONS preflight requests.
+ * 
+ * @param options - CORS configuration options
+ * @returns A middleware function that handles CORS
+ * 
+ * @example
+ * ```typescript
+ * import { cors } from '@aws-lambda-powertools/event-handler/rest';
+ * 
+ * // Use default configuration
+ * app.use(cors());
+ * 
+ * // Custom configuration
+ * app.use(cors({
+ *   origin: 'https://example.com',
+ *   allowMethods: ['GET', 'POST'],
+ *   credentials: true,
+ * }));
+ * 
+ * // Dynamic origin with function
+ * app.use(cors({
+ *   origin: (origin, reqCtx) => {
+ *     const allowedOrigins = ['https://app.com', 'https://admin.app.com'];
+ *     return origin && allowedOrigins.includes(origin);
+ *   }
+ * }));
+ * ```
+ */
+export const cors = (options: CorsOptions = {}): Middleware => {
+  const config = resolveConfiguration(options);
+  
+  return async (_params: Record<string, string>, reqCtx: RequestContext, next: () => Promise<HandlerResponse | void>) => {
+    const { request } = reqCtx;
+    const method = request.method.toUpperCase();
```

Comment:
We can remove `toUpperCase` because method is always capitalised on a `Request` object.